src/datumaro/plugins/data_formats/cityscapes.py

```
@@ -303,8 +303,8 @@ def _load_items(self):
             item_id = self._get_id_from_mask_path(mask_path, mask_suffix)
 
             anns = []
-            instances_mask = load_image(mask_path, dtype=np.int32)
-            segm_ids = np.unique(instances_mask)
+            instances_mask = lazy_image(mask_path, dtype=np.int32)
```

This will work differently from other mask formats, because in other cases item annotations are always immediately available. Have you tested performance of this solution? I think we might need to implement source file caching or RLE recoding to avoid reading from disk if there are more than 2 (default image cache size) mask sources in a single dataset item. If the item image is also used.

---

I used the same script you used in https://github.com/cvat-ai/datumaro/pull/94 but for cityscapes and without streaming to compare performance for annotations attached to the ticket.
Results:
develop: 55.7 seconds, 3.2 Gb
this PR: 55.8 seconds, 174 Mb
Not much difference, but these results could be contaminated by the fact that my machine has almost no RAM to spare and uses swap more extensively on RAM-heavy tasks.

If the lazy reading of images is a problem, how about the following option:
Read images as png (it's always png for this format afaik) but not decode them, and keep them in memory as png bytes. And decode them lazily
It should significantly reduce RAM usage, while keeping previous behaviour of only reading images once

---

> I used the same script ...

I think it not representative enough if unchanged. In that case all the masks are read only once per dataset item, so it basically just moves mask reading to the exporter. It's not bad, but  could you also check the case with several mask sources per dataset item or with several uses of the same loaded mask?

> If the lazy reading of images is a problem, how about the following option ...

I agree, that's exactly what I was talking about above in 

> I think we might need to implement source file caching or RLE recoding

We just need to understand if repeated mask reading can really be a significant performance problem in typical use cases. If we decide that the implemented approach is ok, some other formats should also be updated, including CamVid, Pascal Voc etc. At a glance, it should work fine for single traversal cases (basically, for streaming), so maybe it can be dependent on the mode.

---

Tried to keep png bytes in memory, and it is slower (59s). Apparently, due to lazy_image cache the file reading and decoding is done only twice, but when keeping pngs in memory, there is only one read, but there are 1 decoding per item + 1 decoding per every instance in the mask (so, 3 decodings per item in this dataset). Unless we also implement cache for decoded pngs, it seems like a bad idea.

I believe the typical use case (at least ours) is to import masks and then save them to cvat or transform to polygons, and content of each item's annotations is accessed only once. And the script is actually representative for such use case

Tried the same script, but with 4 exports instead of one (so, 4 uses of masks):
develop: 148.3 s / per item: 1 read, 1 decode
this pr: 164.6 s / per item: 5 reads, 5 decodes
trivial png bytes keeping: 176 s / per item: 1 read, 9 decodes

So, as I see it, there are 3 options
1) how it worked previously: 1 read+decode per item no matter how many times masks are used. But also consumes a lot of memory, so not really an option
2) this pr: 1 read+decode per item on import and 1 read+decode for every use
3) keep pngs in RAM, decode them when they are needed and implement cache, similarly to how it is done in lazy_image, so 1 read+decode per item on import and 1 decode for every use. I doubt it will give significant improvement over (2) since it seems that most of excess time comes from decoding

maybe there is some other small representation which can be kept in memory and decoded more efficiently then png?

---

So far I think we should use implementation from this PR for streaming cases and the old one for other cases. This will allow us to get best out of both of them.

---

Number of decodes is not equal, because I tried very trivial thing, just 
```
+            with open(mask_path, 'rb') as f:
+                image_png_bytes = f.read()
+
+            instances_mask = partial(decode_image, image_png_bytes, dtype=np.int32)
-            instances_mask = lazy_image(mask_path, dtype=np.int32)
```
So lazy_image caches decoded image, and my png-caching-solution does not. Even though lazy_image's cache only works while annotations of the same item are iterated and the moment next item's annotations is accessed cache is emptied, it's still better then no cache at all.

I did not try to implement scheme like (png reading + png cache) -> (png decoding + decoded cache) -> mask.image yet, but I believe it will not be much faster then current lazy_image solution

The huge memory consumption is also a problem for non-streaming cases. With the dataset from the ticket - it's ~750 images, 11 Mb on a disk, and it's 3.2 Gb of memory consumption.
What happens if there are 10 times as many items? It will need 32 Gb RAM to import on develop, or it will need 170 Mb but will take 3 minutes longer on this pr (in the scenario with 4 uses of masks).

---

Made a quick-and-dirty implementation for png-caching + decoded-caching just to be sure
```
@lru_cache(maxsize=None)
def infinitely_caching_png_reader(path):
    with open(path, 'rb') as f:
        return f.read()

@lru_cache(maxsize=1)
def briefly_caching_png_decoder(path):
    return decode_image(infinitely_caching_png_reader(path), dtype=np.int32)
```

```
-            instances_mask = lazy_image(mask_path, dtype=np.int32)
+            instances_mask = partial(briefly_caching_png_decoder, mask_path)
```

And result time is 164.7 s - yes, its not really faster then lazy_image approach

---

Ok, please update other mask formats to this implementation then. Some of them are pascal voc, coco panoptic/stuff (one of the options there), camvid.

---

> What happens if there are 10 times as many items? It will need 32 Gb RAM to import on develop, or it will need 170 Mb but will take 3 minutes longer on this pr (in the scenario with 4 uses of masks).

This question, in particular, makes me think we should add memory limits to the image cache, in addition to the existing count-only approach.

---

In all of these formats it already works in the same way

---

Ok, it seems to be a cityscapes only issue then.